src/cloudWatchLogs/commands/searchLogGroup.ts

```diff
@@ -149,18 +146,95 @@ async function logGroupsToArray(logGroups: AsyncIterableIterator<CloudWatchLogs.
     return logGroupsArray
 }
 
-export function createFilterpatternPrompter(logGroupName: string, isFirst: boolean): InputBoxPrompter {
+/**
+ * HACK: this subclass overrides promptUser() so that we can validate the
+ * search pattern against the service and if it fails, keep the prompt displayed.
+ *
+ * This is necessary until vscode's inputbox.onDidAccept() awaits async callbacks:
+ *    - https://github.com/aws/aws-toolkit-vscode/pull/3114#discussion_r1085484630
+ *    - https://github.com/microsoft/vscode/blob/78947444843f4ebb094e5ab4288360010a293463/extensions/git-base/src/remoteSource.ts#L13
+ *    - https://github.com/microsoft/vscode/blob/78947444843f4ebb094e5ab4288360010a293463/src/vs/base/browser/ui/inputbox/inputBox.ts#L511
+ */
+export class SearchPatternPrompter extends InputBoxPrompter {
+    constructor(
+        public logGroup: CloudWatchLogsGroupInfo,
+        public logParams: CloudWatchLogsParameters,
+        /** HACK: also maintain ad-hoc state because `wizardState` is not mutable. */
+        public readonly retryState: any,
```

Comment:
😬  I tried attaching stuff to `WizardState` but it's immutable (that's probably a good thing :)

I need some way to set `this.recentItem` and `inputBox.validationMessage` (see below).

Reply:
Yeah it's immutable. Overriding `InputBoxPrompter` is a good strategy if you want a lot of customization.